clique/src/metrics.rs

```diff
@@ -1,41 +1,94 @@
-pub trait NetworkCliqueMetrics {
-    fn set_incoming_connections(&self, present: u64);
-    fn set_missing_incoming_connections(&self, missing: u64);
-    fn set_outgoing_connections(&self, present: u64);
-    fn set_missing_outgoing_connections(&self, missing: u64);
+use substrate_prometheus_endpoint::{register, Gauge, PrometheusError, Registry, U64};
+
+#[derive(Clone)]
+pub enum Metrics {
+    Noop,
+    Metrics {
+        incoming_connections: Gauge<U64>,
+        missing_incoming_connections: Gauge<U64>,
+        outgoing_connections: Gauge<U64>,
+        missing_outgoing_connections: Gauge<U64>,
+    },
 }
 
-impl<M: NetworkCliqueMetrics> NetworkCliqueMetrics for Option<M> {
-    fn set_incoming_connections(&self, present: u64) {
-        if let Some(m) = self {
-            m.set_incoming_connections(present)
+impl Metrics {
+    pub fn new(registry: Option<Registry>) -> Result<Self, PrometheusError> {
+        match registry {
+            Some(registry) => Ok(Metrics::Metrics {
+                incoming_connections: register(
+                    Gauge::new(
+                        "clique_network_incoming_connections",
+                        "present incoming connections",
+                    )?,
+                    &registry,
+                )?,
+                missing_incoming_connections: register(
+                    Gauge::new(
+                        "clique_network_missing_incoming_connections",
+                        "difference between expected and present incoming connections",
+                    )?,
+                    &registry,
+                )?,
+                outgoing_connections: register(
+                    Gauge::new(
+                        "clique_network_outgoing_connections",
+                        "present outgoing connections",
+                    )?,
+                    &registry,
+                )?,
+                missing_outgoing_connections: register(
+                    Gauge::new(
+                        "clique_network_missing_outgoing_connections",
+                        "difference between expected and present outgoing connections",
+                    )?,
+                    &registry,
+                )?,
+            }),
+            None => Ok(Metrics::Noop),
         }
     }
 
-    fn set_missing_incoming_connections(&self, missing: u64) {
-        if let Some(m) = self {
-            m.set_missing_incoming_connections(missing)
-        }
+    pub fn noop() -> Self {
+        Metrics::Noop
     }
 
-    fn set_outgoing_connections(&self, present: u64) {
-        if let Some(m) = self {
-            m.set_outgoing_connections(present)
+    pub fn set_incoming_connections(&self, present: u64) {
```

Comment:
We could also use an enum and have one fn set_connections_change:
```rust
Enum XXX {
 IncomingConnections(u64),
 ...,
}
```
names are just for this example pls pick better ones  than mine :P

Reply:
Ugh, I think I'll just do this properly – I left it like that because it's all being called in one place anyway, which is very weird from a metrics perspective. I'll rewrite it so that it actually tracks the metrics live and then the enum makes somewhat more sense.

Reply:
Nice